downstairs/src/region.rs

```diff
@@ -1165,6 +1187,13 @@ impl Extent {
             }
         }
 
+        // We do all of our metadb updates in a single transaction to minimize
+        // syncs.  (Note that the "unchecked" in the signature merely denotes
+        // that we are taking responsibility for assuring that we are not
+        // in a nested transaction, accepting that it will fail at run-time
+        // if we are.)
+        let tx = inner.metadb.unchecked_transaction()?;
```

Comment:
Moving this statement before `set_dirty` is suspect, like others have pointed out, but beyond that I don't understand why moving this statement (or the others in this PR) increases performance at all.

Is it that using `unchecked_transaction` is faster than `transaction`? Does the statement movement have to occur, or is the switch to `unchecked_transaction` that helps?

Reply:
> When the metadb operations are grouped into a transaction, there is a single sync on the commit of the transaction; when they are not in a transaction, every update induces a sync

I get that part, and after rereading my comment think that my question was not getting at my my lack of understanding properly: it doesn't look like moving that statement **should** help, as @faithanalog pointed out: `set_dirty` doesn't accept `tx` as a parameter and use it.

Reading @jgallagher's comment makes this make more sense to me:

> The SQLite C API doesn't have a notion of transactions at all - it just has a connection, and opening/committing/rolling back a transaction all happen via SQL statements executed on that connection.

Reply:
We should change those inner functions to accept `tx` as a parameter - if not in this PR there should be a follow up issue.

Reply:
> Is it that using `unchecked_transaction` is faster than `transaction`?

It's definitely not faster: `transaction` is [just a call to `unchecked_transaction`](https://docs.rs/rusqlite/0.29.0/src/rusqlite/transaction.rs.html#107-109); the difference is in the lifetime on the connection. `transaction()` requires a `&mut Connection` (thus ensuring there are no other open `Transaction`s on this same connection statically, assuming no funny business on the connection that rusqlite doesn't know about), whereas `unchecked_transaction()` accepts `&Connection` (which will allow programs to compile that open nested transactions, which will then fail at runtime).

Reply:
I'll paste what I said in a meeting: alternatively, we should remove `tx` as a parameter, if it's not necessary, and explain that we're using sqlite transactions without needing to worry about `rusqlite::Transaction`.